### ISSUE TITLE
[feature request] time-distributed layers for application of normal layers to sequence data

### ISSUE THREAD

**Issue by erogol** [pytorch/pytorch#1927]
As we move from images to videos, it seems imperative to feed sequential data into common image layers. However, the problem of dealing sequential data with such layers is not clears on Pytorch. I suggest here to devise something like -time-distributed layers for wrapping normal layers and aligning them for sequential inputs and outputs. I think in the course of time, it is going to be a more important pattern to have in Pytorch. Hope, I explain the problem clearly?

Lately, I solved that problem for 1D feat vectors as follows; I wait some comments to generalize it if necessary. 
```
class TimeDistributed(nn.Module):
    def __init__(self, module):
        super(TimeDistributed, self).__init__()
        self.module = module

    def forward(self, x):
        if len(x.size()) <= 2:
            return self.module(x)
        t, n = x.size(0), x.size(1) 
        # merge batch and seq dimensions
        x_reshape = x.contiguous().view(t * n, x.size(2))
        y = self.module(x_reshape)
        # We have to reshape Y
        y = y.contiguous().view(t, n, y.size()[1])
        return y
```

**Comment by pranv** [comment]
I have thought about such an API a lot, both with respect to PyTorch and (earlier) with respect to Keras. Here are my thoughts:

What you've proposed in the code here is an abstraction of what seems to be the ubiquitous pattern used when implementing RNNs. The RNN/LSTM implementation always outputs a 3D tensor and if we want to apply softmax over it, we need to apply the same projection for each samples in the batch, at each time-step. There is a lot of utility for a wrapper that takes in the `Linear` layer and applies it over a 3D tensor. PyTorch's own examples have a couple of . 

But I would argue that this is still a very limited use case and doesn't warrant a generic container. We can have a `LinearOverTime` layer or even extend the `Linear` layer to work with 3D inputs (implicitly flattening across one of the dimensions).

However, there are a lot of use cases where a collection of layers have to be applied repeatedly over an axis. For making such a container for this generic case, we have to consider 2 cases:

1. **Stateful Recurrence**: We have some state that is carried as we index along a particular axis. This cannot be parallelized, but the container can provide some utilities that manage the state transmissions.
2. **Stateless Recurrence**: In this case we can do something similar to what you've proposed -- flatten along some axis (user provided) and do the computation in parallel. (This is for speed only, as stateless recurrence with no state is a strict superset of this).

Common use cases for the first category are attention models, fancy memory augmented networks and novel RNN formulations (ex. Recurrent Highway Networks, Unitary Evolution RNNs etc.,). Common use cases for the second category are the time distributed (or axis distributed) versions of any cascade of layers. 

Basically, I am imagining PyTorch's take on `theano.scan`: http://deeplearning.net/software/theano/library/scan.html . I think it'll be much more user friendly and powerful with PyTorch's dynamic paradigm.

If more people think that this is a good idea, I can work on it :)

**Comment by jekbradbury** [comment]
Here is a generic TimeDistributed mixin (the name for the same idea in Torch7 was `nn.Bottle`):
```
class Bottle(nn.Module):
    def forward(self, input):
        if len(input.size()) <= 2:
            return super(Bottle, self).forward(input)
        size = input.size()[:2]
        out = super(Bottle, self).forward(input.view(size[0]*size[1], -1))
        return out.view(*size, -1)
#example:
class Linear(Bottle, nn.Linear):
    pass
```
That, or a wrapper like the one above, covers your case 2. For your case 1, we've felt so far that using a native Python `for` or `while` loop is the most straightforward and clear way to do this. That's how the decoder in OpenNMT-py is implemented, for instance. Basically, Theano and TensorFlow have to introduce a custom operator like `scan` or `while_loop` because they can't use native Python control flow -- but we can, and there's no reason not to.

**Comment by jekbradbury** [comment]
Meanwhile this https://github.com/pytorch/pytorch/pull/1935 will make TimeDistributed/Bottle unnecessary for `Linear` layers.

**Comment by erogol** [comment]
@jekbradbury but my feeling is (not verified) using only tensor operations without explicit loop would be faster. For example, assume that we have 12x8x3x224x224 input s.t. 8 is seq length. Looping means 8 times layer call, otherwise, if we just merge first two dimensions, this is only one pass on GPU (if memory allows you). And the complexity increases as we increase number of time-distributed layers. Correct me I am wrong.

**Comment by jekbradbury** [comment]
Yes, for the operations that can be parallelized across time or other dimensions you should use tensor operations (or implicit tensor operations in the new `Linear`). But for stateful recurrence you should use a native loop.

**Comment by t-vi** [comment]
Just to add something to this classic issue:

- I'd like a functional wrapper, too.
- Of course it should work with packed sequences (not hard to do either).
- Finally the bikeshedding: Maybe "elementwise" is a more accurate description than time distributed? After all, this throws all timesteps in one giant batch rather than each timestep separately.

I've implemented this a few times now for things like embeddings and more obscure transformations, so I'd love to see this in torch.nn.rnn.utils and as a Module...

If we want this, I'd be very happy to submit a PR.

**Comment by apaszke** [comment]
Wouldn't it be simpler to just implement this as a decorator on the forward?

**Comment by t-vi** [comment]
The functional version is essentially a equivalent to a decorator.
But I'm not sure how I would go using the decorator for an existing class (I used it for Embedding and Bilinear, the latter with two args) without a wrapper or mixin class.

**Comment by ily-R** [comment]
This worked for me : [https://discuss.pytorch.org/t/timedistributed-cnn/51707/2?u=ilyes](https://discuss.pytorch.org/t/timedistributed-cnn/51707/2?u=ilyes)

**Comment by gchanan** [comment]
@t-vi can you clarify if you still think there is something useful to implement here and if so, what it is exactly?  The current description is a bit unclear.

**Comment by t-vi** [comment]
So what the feature request is about is an easy way to take a nn.Module (or a function) that works on some `batch x something` and apply it for each sequence element in on a `seq_len x batch x something` or `batch x seq_len x something` or a packed input.
(See the code above for the functional part.)
I'm not aware that we have that, it seems to come up now and then.

**Comment by t-vi** [comment]
Recent example from the forums: https://discuss.pytorch.org/t/timedistributed-cnn/51707/6
In particular, the code snippet offered in that thread is *not* the equivalent of Keras's TimeDistributed as it duplicates the layers instead of applying the same to all of them.

**Comment by apaszke** [comment]
Maybe we could just provide a function decorator that does this? Then you'd do:
```python
class MyModule(nn.Module):
    ...
    @torch.utils.ndvectorize
    def forward(self, input):
         return ...
```

**Comment by ezyang** [comment]
It seems like there are a few things going on here.

1. With respect to the very first post, there are a bunch of operators which don't support arbitrarily many batch dimensions. When an operator supports arbitrarily many batch dimensions, you don't need `TimeDistributed`, since things will automatically generalize even if you have a batch dimension and a sequence dimension. So one way to make life better here is to just start generalizing operators that only take a single batch dimension to start taking arbitrarily many batch dimensions. (As jekbradbury wrote above, this is indeed what happened #1935) Sometimes you can't conveniently do this though, because that would introduce ambiguity (but generally you do know how many feature dimensions there are, since Modules need to know this up front to allocate parameters anyway)
2. This is all very well and good for modules contributed by PyTorch core, but PyTorch is bigger than the core library, and there is always a place for something like TimeDistributed for massaging third party code that has been implemented under the assumption of only one batch dimension. Various suggestions involving decorators have been put in this thread, but I think @zou3519's work on vmap will thoroughly put this to rest.

**Comment by salimkhazem** [comment]
Hello everyone, I do not know about your progress on this subject, but we (with @jeremyfix) wrote this version of TimeDistributed which can handle dense time tensors as well as PackedSequences. As other recurrent networks, it can also deal with batch first or time first ; Do you think it could be worth a PR ?

```
import torch
import torch.nn as nn
from torch.nn.utils.rnn import PackedSequence
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence


class TimeDistributed(nn.Module):
    """
    TimeDistributed for Pytorch which allows to apply a layer to every temporal slice of an input
    Args:
        Module: a Module instance
    PS : Input must be in the shape of (Seq_length, BS, )
    """

    def __init__(self, module, batch_first=False):
        if not isinstance(module, nn.Module):
            raise ValueError(
                "Please initialize `TimeDistributed` with a "
                f"`torch.nn.Module` instance. Received: {module.type}"
            )
        super(TimeDistributed, self).__init__()
        self.module = module
        self.batch_first = batch_first

    def forward(self, x):
        # Input shape is (default)
        #   - Seq_length, BS, *
        # or if batch_first
        #   - BS, Seq_length, *
        # It can also be provided as PackedSequence
        orig_x = x
        if isinstance(x, PackedSequence):
            x, lens_x = pad_packed_sequence(x, batch_first=self.batch_first)

        if self.batch_first:
            # BS, Seq_length, * -> Seq_length, BS, *
            x = x.transpose(0, 1)
        print(x.shape)
        output = torch.stack([self.module(xt) for xt in x], dim=0)
        if self.batch_first:
            # Seq_length, BS, * -> BS, Seq_length, *
            output = output.transpose(0, 1)

        if isinstance(orig_x, PackedSequence):
            output = pack_padded_sequence(output, lens_x, batch_first=self.batch_first)
        return output
```
`
